### ISSUE TITLE
Options --repo-path and --repo-remote-path ignored in archive-push

### ISSUE THREAD

**Issue by wieck** [pgmasters/backrest#183]
Using verison 0.90 I can't get the archive-push to work unless /etc/pg_backrest.conf contains the correct repo-path in the global:general section. Any --repo-path or --repo-remote-path options are ignored.

Archive command is:

'pg_backrest --stanza=bmsql1_test --backup-host=172.21.21.65 --backup-user=backrest_test "--repo-path=/var/lib/backrest_test" "--repo-remote-path=/var/lib/backrest_test" "--db-path=/var/lib/pgsql/9.4/data" archive-push %p'

Error message is:

ERROR: [122]: remote process terminated: ERROR [122]: unable to create lock path /var/lib/backup/lock

**Comment by dwsteele** [comment]
Hi Jan,

I'm not able to reproduce this with as close a test case as I could manage.  All options are on the command line with no pg_backrest.conf:
```
pg_backrest --backup-host=backup --backup-user=backrest "--repo-path=/var/lib/backrest" "--repo-remote-path=/var/lib/backrest" "--db-path=/var/lib/postgresql/9.4/demo" --log-level-console=debug --stanza=demo archive-push "/var/lib/postgresql/9.4/demo/pg_xlog/000000070000000000000018"
```
It would be helpful if you could add `--log-level-console=debug` to your command line and then attach the output.  I'd be most interested in two lines:
```
INFO: archive-push start: --backup-host=backup --backup-user=backrest --db-path=/var/lib/postgresql/9.4/demo --log-level-console=debug --repo-path=/var/lib/backrest --repo-remote-path=/var/lib/backrest --stanza=demo
<a few log lines>
DEBUG:     Protocol::CommonMaster->new(): iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3, iProtocolTimeout = 1830, strCommand = ssh -o Compression=no -o PasswordAuthentication=no backrest@backup '/usr/bin/pg_backrest --command=archive-push --repo-remote-path=/var/lib/backrest --stanza=demo remote', strName = remote
```
The first line shows all options that were gathered from various sources.  Do you have a pg_backrest.conf file at all?  If so, please attach that as well.  BTW, you can use --no-config on the command line to ignore the conf file entirely and use only command line options.
The second line shows the actual command sent to the remote during protocol start.  This will be helpful for figuring out what's going on.

Thanks!
-David

**Comment by wieck** [comment]
Hi David,

thanks for the fast reply.

On Tue, Apr 5, 2016 at 9:02 AM, David Steele <notifications@github.com>
wrote:

> Hi Jan,
>
> I'm not able to reproduce this with as close a test case as I could
> manage. All options are on the command line with no pg_backrest.conf:
>

Even without a pg_backrest.conf file on both servers (backup and db) I get
the error about /var/lib/backup/lock.

It only works when I create the /var/lib/backup/lock directory with write
permissions for the backrest user. Despite the --repo-path and
--repo-remote-path options, I see the archive-push_remote.lock file
appearing there, not in the specified repo path. Is it possible that
/var/lib/backup exists on your backup server?

The background behind this is that I want to set up a central backup server
with multiple backrest users to serve isolated environments, so I need to
specify the repo paths on the command line.


Regards, Jan


> pg_backrest --backup-host=backup --backup-user=backrest "--repo-path=/var/lib/backrest" "--repo-remote-path=/var/lib/backrest" "--db-path=/var/lib/postgresql/9.4/demo" --log-level-console=debug --stanza=demo archive-push "/var/lib/postgresql/9.4/demo/pg_xlog/000000070000000000000018"
>
> It would be helpful if you could add --log-level-console=debug to your
> command line and then attach the output. I'd be most interested in two
> lines:
>
> INFO: archive-push start: --backup-host=backup --backup-user=backrest --db-path=/var/lib/postgresql/9.4/demo --log-level-console=debug --repo-path=/var/lib/backrest --repo-remote-path=/var/lib/backrest --stanza=demo
> <a few log lines>
> DEBUG:     Protocol::CommonMaster->new(): iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3, iProtocolTimeout = 1830, strCommand = ssh -o Compression=no -o PasswordAuthentication=no backrest@backup '/usr/bin/pg_backrest --command=archive-push --repo-remote-path=/var/lib/backrest --stanza=demo remote', strName = remote
>
> The first line shows all options that were gathered from various sources.
> Do you have a pg_backrest.conf file at all? If so, please attach that as
> well. BTW, you can use --no-config on the command line to ignore the conf
> file entirely and use only command line options.
> The second line shows the actual command sent to the remote during
> protocol start. This will be helpful for figuring out what's going on.
>
> Thanks!
> -David
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pgmasters/backrest/issues/183#issuecomment-205791925>
>



-- 
Jan Wieck
Senior Postgres Architect
http://slony.info

**Comment by dwsteele** [comment]
Hi Jan,

My tests intentionally use a different directory.  So you are saying that the archive logs arrive in the specified repo but it always wants to create the lock in /var/lib/backup?  The still indicates that somehow the preferred repo is not being passed the the remote.  The master process then drives where the data goes and it has the correct remote repo path.

I'll probably need to see the debug info as specified above to be able to diagnose this.

Thanks,
-David

**Comment by wieck** [comment]
Here it is:

2016-04-05 17:08:30.113 T00   INFO: archive-push start:
>> --backup-host=172.21.21.65 --backup-user=backrest_test
>> --db-path=/var/lib/pgsql/9.4/data --log-level-console=debug
>> --repo-path=/var/lib/backrest_test
>> --repo-remote-path=/var/lib/backrest_test --stanza=bmsql1_test
>
> 2016-04-05 17:08:30.113 T00   INFO: push WAL segment
>> pg_xlog/000000280000000E0000000E
>
> 2016-04-05 17:08:30.115 T00  DEBUG:     Archive->pushProcess(): bAsync =
>> false, strSourceFile = pg_xlog/000000280000000E0000000E
>
> 2016-04-05 17:08:30.115 T00  DEBUG:     Protocol::RemoteMaster->new():
>> iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3,
>> iProtocolTimeout = 1830, strCommand = /usr/bin/pg_backrest
>> --command=archive-push --repo-remote-path=/var/lib/backrest_test
>> --stanza=bmsql1_test remote, strHost = 172.21.21.65, strUser = backrest_test
>
> 2016-04-05 17:08:30.115 T00  DEBUG:     Protocol::CommonMaster->new():
>> iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3,
>> iProtocolTimeout = 1830, strCommand = ssh -o Compression=no -o
>> PasswordAuthentication=no backrest_test@172.21.21.65
>> '/usr/bin/pg_backrest --command=archive-push
>> --repo-remote-path=/var/lib/backrest_test --stanza=bmsql1_test remote',
>> strName = remote
>
> 2016-04-05 17:08:30.732 T00  DEBUG:     File->new(): iThreadIdx = [undef],
>> oProtocol = [object], strBackupPath = /var/lib/backrest_test,
>> strDefaultFileMode = <0640>, strDefaultPathMode = <0750>, strRemote =
>> backup, strStanza = bmsql1_test
>
> 2016-04-05 17:08:30.732 T00  DEBUG:     Common:::Lock::lockStopTest():
>> strRepoPath = </var/lib/backrest_test>
>
> 2016-04-05 17:08:30.733 T00  DEBUG:     Archive->walInfo(): strWalFile =
>> /var/lib/pgsql/9.4/data/pg_xlog/000000280000000E0000000E
>
> 2016-04-05 17:08:30.733 T00  DEBUG:     Archive->walInfo=>: strDbVersion =
>> 9.4, ullDbSysId = 6265424879682402960
>
> 2016-04-05 17:08:30.734 T00  DEBUG:     Archive->pushCheck(): bPartial =
>> false, oFile = [object], strDbVersion = 9.4, strWalFile =
>> /var/lib/pgsql/9.4/data/pg_xlog/000000280000000E0000000E, strWalSegment =
>> 000000280000000E0000000E, ullDbSysId = 6265424879682402960
>
> 2016-04-05 17:08:30.739 T00  DEBUG:     Archive->pushCheck=>: strArchiveId
>> = 9.4-1, strChecksum = [undef]
>
> 2016-04-05 17:08:30.740 T00  DEBUG:     File->copy(): bAppendChecksum =
>> true, bDestinationCompress = true, bDestinationPathCreate = true,
>> bIgnoreMissingSource = <false>, bSourceCompressed = false,
>> lModificationTime = [undef], strDestinationFile =
>> 9.4-1/000000280000000E0000000E.gz, strDestinationPathType = backup:archive,
>> strGroup = [undef], strMode = <0640>, strSourceFile =
>> /var/lib/pgsql/9.4/data/pg_xlog/000000280000000E0000000E, strSourcePathType
>> = db:absolute, strUser = [undef]
>
> 2016-04-05 17:08:31.571 T00  DEBUG:     Exit::exitSafe(): iExitCode = 0,
>> strSignal = [undef]
>
> 2016-04-05 17:08:31.571 T00   INFO: archive-push stop
>
> 2016-04-05 17:08:31.571 T00  DEBUG:     Common:::Lock::lockRelease():
>> bFailOnNoLock = false
>
> 2016-04-05 17:08:31.995 T00   INFO: archive-push start:
>> --backup-host=172.21.21.65 --backup-user=backrest_test
>> --db-path=/var/lib/pgsql/9.4/data --log-level-console=debug
>> --repo-path=/var/lib/backrest_test
>> --repo-remote-path=/var/lib/backrest_test --stanza=bmsql1_test
>
>
From the debugging it looks like it is handing down the paths just fine.
/var/lib/backup does not show in any of that output. However, if I delete
the /var/lib/backup directory, this happens:



2016-04-05 17:15:22.186 T00   INFO: archive-push start:
> --backup-host=172.21.21.65 --backup-user=backrest_test
> --db-path=/var/lib/pgsql/9.4/data --log-level-console=debug
> --repo-path=/var/lib/backrest_test
> --repo-remote-path=/var/lib/backrest_test --stanza=bmsql1_test

2016-04-05 17:15:22.186 T00   INFO: push WAL segment
> pg_xlog/000000280000000E00000012

2016-04-05 17:15:22.186 T00  DEBUG:     Archive->pushProcess(): bAsync =
> false, strSourceFile = pg_xlog/000000280000000E00000012

2016-04-05 17:15:22.187 T00  DEBUG:     Protocol::RemoteMaster->new():
> iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3,
> iProtocolTimeout = 1830, strCommand = /usr/bin/pg_backrest
> --command=archive-push --repo-remote-path=/var/lib/backrest_test
> --stanza=bmsql1_test remote, strHost = 172.21.21.65, strUser = backrest_test

2016-04-05 17:15:22.187 T00  DEBUG:     Protocol::CommonMaster->new():
> iBufferMax = 4194304, iCompressLevel = 6, iCompressLevelNetwork = 3,
> iProtocolTimeout = 1830, strCommand = ssh -o Compression=no -o
> PasswordAuthentication=no backrest_test@172.21.21.65
> '/usr/bin/pg_backrest --command=archive-push
> --repo-remote-path=/var/lib/backrest_test --stanza=bmsql1_test remote',
> strName = remote

2016-04-05 17:15:22.741 T00  DEBUG:     File->new(): iThreadIdx = [undef],
> oProtocol = [object], strBackupPath = /var/lib/backrest_test,
> strDefaultFileMode = <0640>, strDefaultPathMode = <0750>, strRemote =
> backup, strStanza = bmsql1_test

2016-04-05 17:15:22.741 T00  DEBUG:     Common:::Lock::lockStopTest():
> strRepoPath = </var/lib/backrest_test>

2016-04-05 17:15:22.742 T00  DEBUG:     Archive->walInfo(): strWalFile =
> /var/lib/pgsql/9.4/data/pg_xlog/000000280000000E00000012

2016-04-05 17:15:22.742 T00  DEBUG:     Archive->walInfo=>: strDbVersion =
> 9.4, ullDbSysId = 6265424879682402960

2016-04-05 17:15:22.742 T00  DEBUG:     Archive->pushCheck(): bPartial =
> false, oFile = [object], strDbVersion = 9.4, strWalFile =
> /var/lib/pgsql/9.4/data/pg_xlog/000000280000000E00000012, strWalSegment =
> 000000280000000E00000012, ullDbSysId = 6265424879682402960

2016-04-05 17:15:22.781 T00  ERROR: [122]: remote process terminated: ERROR
> [122]: unable to create lock path /var/lib/backup/lock

2016-04-05 17:15:22.783 T00  DEBUG:     Exit::exitSafe(): iExitCode = 122,
> strSignal = [undef]

2016-04-05 17:15:22.783 T00   INFO: archive-push stop



Regards, Jan


On Tue, Apr 5, 2016 at 12:18 PM, David Steele <notifications@github.com>
wrote:

> Hi Jan,
>
> My tests intentionally use a different directory. So you are saying that
> the archive logs arrive in the specified repo but it always wants to create
> the lock in /var/lib/backup? The still indicates that somehow the preferred
> repo is not being passed the the remote. The master process then drives
> where the data goes and it has the correct remote repo path.
>
> I'll probably need to see the debug info as specified above to be able to
> diagnose this.
>
> Thanks,
> -David
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pgmasters/backrest/issues/183#issuecomment-205878976>
>



-- 
Jan Wieck
Senior Postgres Architect
http://slony.info

**Comment by dwsteele** [comment]
Hi Jan,

Thanks for the additional information.  As it turned out I had failed to remove pg_backrest.conf on the backup server and that's why I was not able to reproduce the error originally.  Getting the debug output narrowed my options and helped me figure out where I had gone wrong.

The fix is on the dev-repofix branch and here's the diff:

https://github.com/pgmasters/backrest/compare/dev-repofix?expand=1

It looks like you are testing rather than running in production so I thought I would give you the chance to try it out before I do a release.  If I got that wrong and you are in production then I'll be happy to release this as 0.92 any time.

Thanks,
-David

**Comment by wieck** [comment]
Hi David,

thank you for the fast turnaround.

On Tue, Apr 5, 2016 at 6:35 PM, David Steele <notifications@github.com>
wrote:

> Hi Jan,
>
> Thanks for the additional information. As it turned out I had failed to
> remove pg_backrest.conf on the backup server and that's why I was not able
> to reproduce the error originally. Getting the debug output narrowed my
> options and helped me figure out where I had gone wrong.
>
> The fix is on the dev-repofix branch and here's the diff:
>
> https://github.com/pgmasters/backrest/compare/dev-repofix?expand=1
>

That branch does everything I tested right. My setup is that on the
dedicated backup server there are two backrest users (there could be more),

backrest HOME=/var/lib/backrest
backrest_test HOME=/var/lib/backrest_test

backrest@<backup-server> is sharing ssh keys with postgres@<db-servers> and
backrest_test@<backup-server> shares with postgres@<db_test-servers>.

On the database servers all repos are owned by postgres and located at
/var/lib/pgsql/backrest.

None of the servers has an /etc/pg_backrest.conf. With that fix this all
provides a shared backup host for otherwise isolated environments.


> It looks like you are testing rather than running in production so I
> thought I would give you the chance to try it out before I do a release. If
> I got that wrong and you are in production then I'll be happy to release
> this as 0.92 any time.
>

It is test at this moment. So far I have tested the archive-get,
archive-push, backup and restore commands and everything is working fine.
How far out is 0.95?


Thanks, Jan

Thanks,
> -David
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly or view it on GitHub
> <https://github.com/pgmasters/backrest/issues/183#issuecomment-206016684>
>



-- 
Jan Wieck
Senior Postgres Architect
http://slony.info

**Comment by dwsteele** [comment]
Hi Jan ,

> thank you for the fast turnaround.

No problem - I know how frustrating it can be to have a bug that's preventing you from getting your desired result.

> <...> So far I have tested the archive-get,
> archive-push, backup and restore commands and everything is working fine.

Glad to hear it!

> How far out is 0.95?

That's hard to say.  I'm currently working on the 1.0 release (see numbered dev branches) and plan to have that ready for my talk in NYC in a few weeks.  Or it might become 0.95 if I don't feel completely comfortable with it.

I'm going to release this fix as 0.92.  If somebody else hits the issue they might just get frustrated and give up rather than engaging me or checking the issues.

Thanks,
-David

**Comment by wieck** [comment]
On Wed, Apr 6, 2016 at 8:43 AM, David Steele <notifications@github.com>
wrote:

>
> How far out is 0.95?
>
> That's hard to say. I'm currently working on the 1.0 release (see numbered
> dev branches) and plan to have that ready for my talk in NYC in a few
> weeks. Or it might become 0.95 if I don't feel completely comfortable with
> it.
>
> I'm going to release this fix as 0.92. If somebody else hits the issue
> they might just get frustrated and give up rather than engaging me or
> checking the issues.
>
>
In that case a 0.92 is definitely appreciated. Much better than installing
a git dev-branch checkout for production, which could happen within days
rather than weeks here.


Thanks a lot, Jan


-- 
Jan Wieck
Senior Postgres Architect
http://slony.info

**Comment by dwsteele** [comment]
Definitely, and here it is:
https://github.com/pgmasters/backrest/releases/tag/release/0.92
